libfoundation/src/foundation-number.cpp

```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+{
+    kMCNumberBinaryTypeSignedIntegerBySignedInteger = (kMCNumberFlagSignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagSignedIntegerRep,
+    kMCNumberBinaryTypeSignedIntegerByUnsignedInteger = (kMCNumberFlagSignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagUnsignedIntegerRep,
+    kMCNumberBinaryTypeSignedIntegerByDouble = (kMCNumberFlagSignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagDoubleRep,
+    kMCNumberBinaryTypeUnsignedIntegerBySignedInteger = (kMCNumberFlagUnsignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagSignedIntegerRep,
+    kMCNumberBinaryTypeUnsignedIntegerByUnsignedInteger = (kMCNumberFlagUnsignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagUnsignedIntegerRep,
+    kMCNumberBinaryTypeUnsignedIntegerByDouble = (kMCNumberFlagUnsignedIntegerRep << kMCNumberFlagRepBitCount) | kMCNumberFlagDoubleRep,
+    kMCNumberBinaryTypeDoubleBySignedInteger = (kMCNumberFlagDoubleRep << kMCNumberFlagRepBitCount) | kMCNumberFlagSignedIntegerRep,
+    kMCNumberBinaryTypeDoubleByUnsignedInteger = (kMCNumberFlagDoubleRep << kMCNumberFlagRepBitCount) | kMCNumberFlagUnsignedIntegerRep,
+    kMCNumberBinaryTypeDoubleByDouble = (kMCNumberFlagDoubleRep << kMCNumberFlagRepBitCount) | kMCNumberFlagDoubleRep,
+};
+
+static inline intenum_t __MCNumberComputeBinaryType(MCNumberRef x, MCNumberRef y)
+{
+    return ((x -> flags & kMCNumberFlagRepMask) << kMCNumberFlagRepBitCount) | (y -> flags & kMCNumberFlagRepMask);
+}
+
+//////////
+
+// This type must be big enough to be able to hold the result of any arithmetic
+// binary operation performed on a combination of integer_t and uinteger_t.
+typedef int64_t bigint_t;
+
+struct __MCNumberOperationAdd
+{
+    static inline double double_by_double(double x, double y)
+    {
+        return x + y;
+    }
+ 
+    static inline bigint_t bigint_by_bigint(bigint_t x, bigint_t y)
+    {
+        return x + y;
+    }
+    
+    static inline bool signed_by_signed(integer_t x, integer_t y, integer_t& r_value)
+    {
+        return __checked_signed_add(x, y, &r_value);
+    }
+    
+    static inline bool unsigned_by_unsigned(uinteger_t x, uinteger_t y, uinteger_t& r_value)
+    {
+        return __checked_unsigned_add(x, y, &r_value);
+    }
+    
+    static inline bool signed_by_unsigned(integer_t x, uinteger_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that y will be in the range (INT_MAX, UINT_MAX]
+        // if x is positive, we cannot produce an integer_t.
+        if (x > 0)
+            return false;
+        
+        // if x <= 0 then
+        //   x + y === -(-x) + y === y - (-x)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(y, (uinteger_t)-x, &z))
+            return false;
+        
+        if (z > INT32_MAX)
+            return false;
+        
+        r_value = (integer_t)z;
+        
+        return true;
+    }
+    
+    static inline bool unsigned_by_signed(uinteger_t x, integer_t y, integer_t& r_value)
```

Comment:
However... it would be nice for this to work (pseudocode):

    variable tUnsigned as unsigned
    variable tNumber
    put INTEGER_MAX into tUnsigned
    put 2 into tNumber
    add tNumber to tUnsigned